Userland/Utilities/ls.cpp

```diff
@@ -142,6 +143,22 @@ int main(int argc, char** argv)
     for (size_t i = 0; i < paths.size(); i++) {
         auto path = paths[i];
 
+        if (flag_recursive && Core::File::is_directory(path)) {
+            size_t subdirs = 0;
+            Core::DirIterator di(path, Core::DirIterator::SkipParentAndBaseDir);
+
+            if (di.has_error()) {
+                status = 1;
+                fprintf(stderr, "%s: %s\n", path, di.error_string());
+            }
+
+            while (di.has_next()) {
+                const String directory = di.next_full_path();
+                if (Core::File::is_directory(directory))
+                    paths.insert(i + (++subdirs), strdup(directory.characters()));
```

Comment:
Let's not leak memory here.
Instead, please switch `paths` to a `Vector<String>` and append `directory` here.
Also, `i + (++subdirs)` is a little complex, how about splitting it up?

Reply:
I changed `paths` to a Vector<String> and used `move()` instead. However, I continue to `insert()` rather than `append()` the directory in order to maintain the order of directories. With `append()`, a directory's subdirectories are not necessarily neighboring it in the `paths`.

As a side note, are there tools in Serenity I can use to check if something I've written is leaking memory? Because I'm not sure I've solved the issue your referring. I'm aware of Valgrind, but I don't know how to use it in this workflow.

Reply:
Oh `insert()` is fine, my bad.
> As a side note, are there tools in Serenity I can use to check if something I've written is leaking memory?

Yes, run it with `UserspaceEmulator` (`ue` in the shell), that's our equivalent of valgrind.